src/ExactDiagonalization/basis_fock.jl

```diff
@@ -0,0 +1,181 @@
+"""
+    build_basis(addr::AbstractFockAddress)
+    build_basis(::Type{<:AbstractFockAddress}) -> basis
+
+Return all possible Fock states of a given type as a vector. This method is _much_ faster
+than `build_basis(::AbstractHamiltonian, ...)`, but does not take matrix blocking into
+account. This version of `build_basis` accepts no additional arguments.
+
+All address types except [`OccupationNumberFS`](@ref Main.Rimu.OccupationNumberFS) are
+supported.
+
+Returns a sorted vector of length equal to the [`dimension`](@ref) of `addr`.
+"""
+function build_basis(addr::AbstractFockAddress)
+    return build_basis(typeof(addr))
+end
+
+###
+### BoseFS
+###
+# Algorithm:
+# Given an address type of N particles in M modes,
+# For all n_1 ∈ (0 ... N)
+# - put n_1 particles in the first mode
+# - recursively call the function to put the remaining (N-n_1) particles into (M-1) modes.
+# The recursion is parallelised through spawning tasks.
+
+# this is equivalent to `dimension(BoseFS{n,m})`, but does not return a `BigInt`.
+_bose_dimension(n, m) = binomial(n + m - 1, n)
+
+function build_basis(::Type{<:BoseFS{N,M}}) where {N,M}
+    T = typeof(near_uniform(BoseFS{N,M}))
+    result = Vector{T}(undef, _bose_dimension(N, M))
+    _bose_basis!(result, (), 1, N, Val(M), 2 * Threads.nthreads())
+    return result
+end
+
+# Multithreaded version - attempts to spawn `n_tasks` tasks.
+@inline function _bose_basis!(
+    result::Vector, postfix, index, remaining_n, ::Val{M}, n_tasks::Int
+) where {M}
+    @sync if M < 5 || remaining_n ≤ 1 || n_tasks ≤ 0
+        _bose_basis!(result, postfix, index, remaining_n, Val(M))
+    else
+        n_tasks ÷= 2
+        Threads.@spawn begin
+            _bose_basis!(result, $(0, postfix...), $index, $remaining_n, Val(M-1), $n_tasks)
+        end
+        index += _bose_dimension(remaining_n, M - 1)
+        Threads.@spawn begin
+            _bose_basis!(
+                result, $(1, postfix...), $index, $(remaining_n - 1), Val(M-1), $n_tasks
+            )
+        end
+        index += _bose_dimension(remaining_n - 1, M - 1)
+        for n in 2:remaining_n
```

Comment:
Why is the serial loop for cases `2:remaining_n` while `(1, postfix)` cases are handled with parallelisation?

Reply:
The idea is to spawn threads on subaddresses with large `remaining_n` as there is more work to do there. Also spawning two threads per iteration makes the number of spawned tasks easier to predict. Not 100% this is the optimal choice but it seems to work well enough.